mixer/adapter/kubernetes2/kubernetes.go

```diff
@@ -0,0 +1,441 @@
+
+	// cache invalidation
+	// TODO: determine a reasonable default
+	defaultRefreshPeriod = 5 * time.Minute
+)
+
+var (
+	conf = &config.Params{
+		KubeconfigPath:                        "",
+		CacheRefreshDuration:                  defaultRefreshPeriod,
+		PodLabelForService:                    podServiceLabel,
+		PodLabelForIstioComponentService:      istioPodServiceLabel,
+		FullyQualifiedIstioIngressServiceName: istioIngressSvc,
+		LookupIngressSourceAndOriginValues:    lookupIngressSourceAndOriginValues,
+		ClusterDomainName:                     clusterDomain,
+	}
+)
+
+type (
+	builder struct {
+		adapterConfig        *config.Params
+		newCacheControllerFn controllerFactoryFn
+		needsCacheInit       bool
+	}
+
+	handler struct {
+		pods   cacheController
+		log    adapter.Logger
+		params *config.Params
+	}
+
+	// used strictly for testing purposes
+	controllerFactoryFn func(kubeconfigPath string, refreshDuration time.Duration, env adapter.Env) (cacheController, error)
+)
+
+// compile-time validation
+var _ kubernetes_apa_tmpl.Handler = &handler{}
+var _ kubernetes_apa_tmpl.HandlerBuilder = &builder{}
+
+/////////////////////////////// Config-time methods //////////////////////////////
+//
+//
+
+// GetInfo returns the Info associated with this adapter implementation.
+func GetInfo() adapter.Info {
+	return adapter.Info{
+		Name:        "KubernetesAttributeGenerator",
+		Impl:        "istio.io/istio/mixer/adapter/kubernetes",
+		Description: "Provides platform specific functionality for the kubernetes environment",
+		SupportedTemplates: []string{
+			kubernetes_apa_tmpl.TemplateName,
+		},
+		DefaultConfig: conf,
+
+		NewBuilder: func() adapter.HandlerBuilder { return newBuilder(newCacheFromConfig) },
+	}
+}
+
+func (b *builder) SetAdapterConfig(c adapter.Config) {
+	b.adapterConfig = c.(*config.Params)
+}
+
+// Validate is responsible for ensuring that all the configuration state given to the builder is
+// correct.
+func (b *builder) Validate() (ce *adapter.ConfigErrors) {
+	params := b.adapterConfig
+	if len(params.PodLabelForService) == 0 {
+		ce = ce.Appendf("podLabelForService", "field must be populated")
+	}
+	if len(params.PodLabelForIstioComponentService) == 0 {
+		ce = ce.Appendf("podLabelForIstioComponentService", "field must be populated")
+	}
+	if len(params.FullyQualifiedIstioIngressServiceName) == 0 {
+		ce = ce.Appendf("fullyQualifiedIstioIngressServiceName", "field must be populated")
+	}
+	if len(params.ClusterDomainName) == 0 {
+		ce = ce.Appendf("clusterDomainName", "field must be populated")
+	} else if len(strings.Split(params.ClusterDomainName, ".")) != 3 {
+		ce = ce.Appendf("clusterDomainName", "must have three segments, separated by '.' ('svc.cluster.local', for example)")
+	}
+	return
+}
+
+func (b *builder) Build(ctx context.Context, env adapter.Env) (adapter.Handler, error) {
+	paramsProto := b.adapterConfig
+	stopChan := make(chan struct{})
+	var pods cacheController
```

Comment:
This would be wrong. Second run of `Build` will leave this `pod` to be `nil`.
Let's add a test case of building multiple handlers from a single builder.

Reply:
Awesome catch. thanks.. added a test and removed the `needsCacheInit = false` code. WDYT ?

Reply:
It's even better to remove `needsCacheInit` itself if it always init the cache

Reply:
yup.. D